### ISSUE TITLE
-Wint-to-void-pointer-cast not triggered by C++ casts

### ISSUE THREAD

**Issue by arichardson** [llvm/llvm-project#53964]
The following code triggers `-Wint-to-void-pointer-cast` in the first function but not the second (
https://godbolt.org/z/G6r8GfMoe):
```c++
void* cstyle(short value) {
    return (void*)value;
}

void* reinterpret(short value) {
    return reinterpret_cast<void*>(value);
}
```

**Comment by LebedevRI** [comment]
That's most likely intentional, that's one of the defined use-cases of `reinterpret_cast<>()`.

**Comment by dwblaikie** [comment]
Looks like GCC's `-Wint-to-pointer-cast` has the same behavior: https://godbolt.org/z/o89dE8YPP 
Might be justified since reinterpret_cast is more explicit about the user's intent.

**Comment by arichardson** [comment]
I see the following comment was added in the commit that introduced the warning (https://github.com/llvm/llvm-project/commit/282ad876bd89a7d7ce698ab4128dc8f0a18105cd):
```
  // Not warning on reinterpret_cast, boolean, constant expressions, etc
  // are not explicit design choices, but consistent with GCC's behavior.
  // Feel free to modify them if you've reason/evidence for an alternative.
```
So it appears this is intended behaviour to match GCC. I was just surprised that the warning behaviour changed when using C++ casts rather than C-style ones, but since this is intended rather than accidental because of different codepaths SemaCast.cpp, I'll close this as "not a bug".
